### DISCUSSION TITLE
Recording Power connections

### DISCUSSION THREAD

**Discussion by sscotter** [netbox-community/netbox#12912] [Help Wanted!]
Hi,
I'm struggling to understand how best way to record the power connections between our servers and PDU rails in the racks.

**Environment**
I have three racks across two datacentres. Each rack has two separate PDUs which (as I understand it) and feed by two different sources and protected by two different diesel generators in the event of a problem. ie, I have six PDU's in total.

I've tried the following strategies

**Strategy 1**
I created two _Power Panels_, one for each datacentre.
I then created six _Power Feeds_ to represent the PDUs in the racks and assigned those to the correct _Power Panel_
* DC 1 - Rack A - Left PDU
* DC 1 - Rack A - Right PDU
* DC 1 - Rack B - Left PDU
* DC 1 - Rack B - Right PDU
* DC 2 - Rack C - Left PDU
* DC 2 - Rack C - Right PDU

This allowed me to record that (for example) Server A in DC 1 has PSU1 connected to _DC 1 - Rack A - Left PDU_ and PSU2 connected to _DC 1 - Rack A - Right PDU_. However, when it came time to record Server B's (also in Rack A) power connectivity, both _DC 1 - Rack A - Left PDU_, and _DC 1 - Rack A - Right PDU_ where greyed out.

At this point it dawned on me that maybe _Power Feeds_ represent the individual plugs on the PDU. 

**Strategy 2**
Deleted all _Power Panels_ and _Power Feeds_ and started again.
Created four _Power Panels_, two for each datacentre.
* Power Panel DC 1 - Left PDUs
* Power Panel DC 1 - Right PDUs
* Power Panel DC 2 - Left PDUs
* Power Panel DC 2 - Right PDUs

I then went to create 16 _Power Feeds_ for _Power Panel DC 1 - Left PDUs_, with the intention of giving them names such as _Rack A - Left PDU - Port [0-15]_ and it literally created ONE _Power feed_ named `Rack A - Left PDU - Port [0-15]`, not 16 _power feeds_ named `Rack A - Left PDU - Port 0`,`Rack A - Left PDU - Port 1` etc. ie, the _Add a new power feed_ page doesn't support bulk creation like most creation pages seem to in Netbox.

Manually creating the following four ports allowed me to record Server A and Server B's power connections. Result!, so now I just need to quickly and accurately create the induvial _Power Feeds_
* Rack A - Left PDU - Port 1
* Rack A - Right PDU - Port 1
* Rack A - Left PDU - Port 2
* Rack A - Right PDU - Port 2

**Strategy 3**
To save myself to finger effort I thought I'd import a CSV with the relevant information PDU names in so they're created consistently and less prone to typos. Using the following CSV to import Power Feeds...

```
site,name,power_panel,rack,status,type,supply,voltage,amperage,phase,max_utilization
Site A,Rack A - Left PDU - Port 1,Power Panel DC 1 - Left PDUs,r203,active,primary,ac,240,20,single-phase,80
```
... I get the following error...
`Record 1 rack: Object not found: R203`.

The rack name is definitely R203. The slug is r203 and the id of the rack is 23. Using either the slug or the id in the CSV results in exactly the same error `Record 1 rack: Object not found: r203` or `Record 1 rack: Object not found: 23`.

It's worth pointing out I don't know why `site` is a required attribute when bulk importing power feeds because it is not a field asked for at all (required or not) when creating a _power feed_ manually.

Am I doing something wrong (or even stupid) or does netbox have a bug in the import of power feeds?

Cheers

Steve

**Comment by candlerb** [comment]
> At this point it dawned on me that maybe Power Feeds represent the individual plugs on the PDU.

They represent the individual feeds *into* the PDU.  Think of a power feed as a circuit breaker that supplies power to you. A power panel is just a grouping where power feeds are located together.

In a typical data centre, there would be two power feeds into each rack - an A and B feed - and Netbox allows each power feed to be optionally associated with a rack (it's mandatory to associate it with a power panel though). Then you'd connect each power feed into a PDU device (to its power port), and then connect the power outlets on the PDU device to the power ports of the other devices in the rack.

I don't fully understand what you've modelled, because you mention PDUs but haven't said how you've connected them. Normally a PDU would have one power port (inlet) and multiple power outlets linked internally to that power port, and then the power outlets would connect to power ports on your servers. (For a similar recent case see #12902)

For your CSV: to import power feeds, if you want to associate them with a rack, you need to use the *name* of the rack. It doesn't matter if the slug is r203 or the database id is 23, it must have *name* of (say) "R203", and I presume must be in the same site.  (Slugs are mostly unused in Netbox, although there are cases you can use them for query filtering in the REST API). Similarly, the name of a rack is only unique within a Location (a.k.a. "rack group" originally), so you may need to specify both the Location Name and the Rack Name to identify it uniquely.

> It's worth pointing out I don't know why `site` is a required attribute when bulk importing power feeds

I believe it's because the name of a power panel is only unique within a site, so the combination of (site_name, power_panel_name) is needed to identify it.

Aside: I've been trying to document the power features in Netbox in more detail, especially with regards how it calculates overall power utilization from individual device power draw. The gory details are in #12837.

**Reply by CvR42** [reply]
A bit late but this seems a reasonable follow-up question:

Power feeds as something that has a circuit breaker makes absolute sense to me, but then why is there the restriction to allow only one connection to a power feed? Real-world wiring normally doesn't have that restriction. Are we supposed to split the power feed explicitly with some kind of PDU that feeds the PDUs?

**Comment by sebmuc99** [comment]
## Residential/Home Use Case: One Circuit Breaker → Multiple Devices

I'm encountering the same limitation mentioned by @CvR42, but in a residential/home lab context rather than a datacenter. The current one-to-one relationship between power feeds and device power ports doesn't align with real-world residential electrical installations.

### The Real-World Scenario

In residential electrical systems, a single circuit breaker (power feed) typically powers **multiple wall outlets**, and each wall outlet can power **multiple devices** (either directly or via power strips). For example:

**Circuit Breaker: F12-Kitchen-Outlets (16A)**
- Powers: KI-OL-01 (Wall Outlet Kitchen 01)
- Powers: KI-OL-02 (Wall Outlet Kitchen 02)  
- Powers: KI-OL-03 (Wall Outlet Kitchen 03)
- Powers: KI-OL-04 (Wall Outlet Kitchen 04)

Each wall outlet then powers one or more devices:
- KI-OL-01 → Refrigerator
- KI-OL-02 → Power Strip → Coffee Maker + Toaster
- KI-OL-03 → Dishwasher
- KI-OL-04 → Microwave

### The Problem

NetBox currently allows a power feed to connect to only ONE device's power port. This means I can only connect F12-Kitchen-Outlets to KI-OL-01, and the feed becomes unavailable for KI-OL-02, KI-OL-03, and KI-OL-04. This doesn't reflect reality - the circuit breaker physically supplies power to all four outlets simultaneously.

### Current Workaround (Doesn't Make Sense)

The only workaround would be to create separate power feeds for each outlet:
- F12-Kitchen-Outlets-Port1
- F12-Kitchen-Outlets-Port2
- F12-Kitchen-Outlets-Port3
- F12-Kitchen-Outlets-Port4

But this is incorrect modeling because:
1. There's only ONE physical circuit breaker, not four
2. All four outlets share the same 16A capacity (not 16A each)
3. Power utilization tracking would be wrong - the circuit could be overloaded even if each "feed" shows under 80%

### Proposed Solution

Power feeds should support **one-to-many connections** to device power ports, similar to how:
- A network switch port can connect to multiple VLANs
- A device can have multiple cables to different ports

This would allow accurate modeling of:
- **Residential installations**: One circuit breaker → multiple wall outlets → devices
- **Industrial panels**: One circuit breaker → multiple machines in a work area
- **Datacenter scenarios**: Main circuit breaker → distribution panel → multiple rack PDUs

### Benefits

1. **Accurate topology**: Reflects real electrical distribution
2. **Proper power tracking**: Total load on circuit breaker = sum of all connected devices
3. **Better capacity planning**: Shows when a circuit is approaching its 16A limit across all outlets
4. **Realistic modeling**: Matches how electricians wire buildings

The datacenter model (one feed → one PDU → multiple outlets) works for that use case, but home/residential installations are structured differently and equally valid to model in NetBox.

Would appreciate hearing thoughts on whether this use case makes sense and if there's potential to support one-to-many power feed relationships in the future.

---

**Example hierarchy for clarity:**
```
Power Panel (Apartment Distribution Panel)
  └─ Power Feed: F12-Kitchen-Outlets (16A Circuit Breaker)
      ├─ KI-OL-01 Power Port → KI-OL-01 Outlet → Refrigerator
      ├─ KI-OL-02 Power Port → KI-OL-02 Outlet → Power Strip → 3 devices
      ├─ KI-OL-03 Power Port → KI-OL-03 Outlet → Dishwasher
      └─ KI-OL-04 Power Port → KI-OL-04 Outlet → Microwave
      
Total load on F12-Kitchen-Outlets = sum of all 4+ devices
```
